### ISSUE TITLE
Cast array or object property as json

### ISSUE THREAD

**Issue by ahocquard** [matthewfranglen/postgres-elasticsearch-fdw#6]
Hello,

Is it possible to cast an array as json easily with the extension? Same thing for json objects.
As far I know and by reading the Python code of the extension, it casts it as a Python string.

```
        {
            "browser_name": "Chrome",
            "user_os_name": "Mac OS X",
            "email_domains": ["example.com", "example2.com"],
            "geoip": {
                "ip": "203.168.53.187",
                "latitude": -37.734,
                "location": {
                    "lat": -37.734,
                    "lon": 144.7378
                }
            }
        }
```

```
select email_domains from my_foreign_table;

[u'akeneo.com', u'example.com']
```

```
select geoip from my_foreign_table;

"region_code"=>"VIC", "longitude"=>144.7378, "region_name"=>"Victoria", "latitude"=>-37.734, "ip"=>"203.168.53.187", "continent_code"=>"OC", "country_code3"=>"AU", "country_code2"=>"AU", "postal_code"=>"3023", "location"=>"lat"=>-37.734, "lon"=>144.7378, "country_name"=>"Australia", "city_name"=>"Caroline Springs", "timezone"=>"Australia/Melbourne"
```

Is it supported?

**Comment by matthewfranglen** [comment]
Hi,

Thanks for opening this issue. This sounds like a good idea.

The code performs no explicit coercion of the data from the elastic search query, however as you have noticed it gets turned into varying types (in your example it is array->array and dictionary->hstore). When I looked up this in the multicorn I found [this issue](https://github.com/Segfault-Inc/Multicorn/issues/86) which confirms the behaviour you have experienced.

I think that the best thing I can do quickly is to ensure that the string representation of the json is always returned, which can then be cast into json or jsonb. For your specific use case it may require a view over the foreign table to apply a cast (when I've implemented this change I can test if this is required).

What do you think?

**Comment by matthewfranglen** [comment]
I also wonder what behaviour would be appropriate when writing to such a column.

**Comment by ahocquard** [comment]
> I think that the best thing I can do quickly is to ensure that the string representation of the json is always returned, which can then be cast into json or jsonb

That would be very handy indeed.

In my case, if a `JSON` type does not work natively,  I can still do:
```
CREATE TABLE IF NOT EXISTS access_log AS SELECT *, domain_names::jsonb as json_domain_names FROM es_access_log;
```
And if declaring my column as `jsonb` instead of `text` works automatically, great!

**Comment by matthewfranglen** [comment]
With regard to writes - it's likely that the data provided via the multicorn library is always a string. That's something that I need to test. The elasticsearch instance would refuse anything that was not a completely formed json object consistent with the schema. Since my original use case for this was flat documents where columns were simple values this was not a problem. Ensuring that the write of a structured column is converted appropriately is the concern.

With regard to the casting of the output column I would remind you that there is no caching going on here. You would be paying the overhead of conversion to jsonb every time. I would recommend testing the performance of the two types carefully. A materialized view might help if you need to operate over the blob.

I should be able to work on this a bit tonight.

**Comment by ahocquard** [comment]
> With regard to the casting of the output column I would remind you that there is no caching going on here. You would be paying the overhead of conversion to jsonb every time. I would recommend testing the performance of the two types carefully. A materialized view might help if you need to operate over the blob.

Yes, casting is done at runtime which is a barrier optimization in many cases. But in my example, I added a new `jsonb` column to convert the `text` column on the fly when importing data from ES. Then, in Postgres I can use the `jsonb` column instead of the TEXT column. I only pay the cost of the cast at import time, which is , I think, negligible. 

Anyway, performance is not a big concern for my usecase. 

> I should be able to work on this a bit tonight.
Thanks for the reactivity! If I can help somehow, tell me.

**Comment by matthewfranglen** [comment]
Oh this is a really great idea. I've created a [jsonb](https://github.com/matthewfranglen/postgres-elasticsearch-fdw/tree/jsonb) branch for you to test out properly. My initial evaluation shows that for reading it works with the json/jsonb type applied to the foreign table directly.

I've listed instructions to reproduce my test run on [the last commit](https://github.com/matthewfranglen/postgres-elasticsearch-fdw/commit/bf66f608b4924b6a9ca2414aa04d367351929bd6). It would be good if you could check it and report any issues you find.

**Comment by ahocquard** [comment]
Hello,

I launched my testsuite successfully with your branch! I had good hope as `timestamp` and `inet` types were automatically casted.
Also, it's possible to generate a valid json this way now (subquery mandatory):

```
select row_to_json(access_log) from (select * from es_access_log) access_log;
```

Thanks!

**Comment by matthewfranglen** [comment]
That's great :smile:

I need to add proper tests around this, implement the write behaviour and document it all. After that I'll release a new version. I'm hoping to get that done over the easter weekend.

**Comment by ahocquard** [comment]
Cool! Thanks a lot for this improvement :)

**Comment by matthewfranglen** [comment]
I've completed the test/documentation updates and released the new version. If you feel that this ticket is resolved then please close it.

**Comment by ahocquard** [comment]
Thanks for the new version (and for the support of Postgres 12). It works as expected!

**Comment by ahocquard** [comment]
Thanks for the new version (and for the support of Postgres 12). It works as expected!

**Comment by matthewfranglen** [comment]
Great stuff :smiley_cat:
